scene/gui/rich_text_label.cpp

```diff
@@ -4173,16 +4177,20 @@ void RichTextLabel::_bind_methods() {
 }
 
 void RichTextLabel::set_visible_characters(int p_visible) {
-	visible_characters = p_visible;
-	if (p_visible == -1) {
-		percent_visible = 1;
-	} else {
-		int total_char_count = get_total_character_count();
-		if (total_char_count > 0) {
-			percent_visible = (float)p_visible / (float)total_char_count;
+	if (visible_characters != p_visible) {
+		visible_characters = p_visible;
+		if (p_visible == -1) {
+			percent_visible = 1;
+		} else {
+			int total_char_count = get_total_character_count();
+			if (total_char_count > 0) {
+				percent_visible = (float)p_visible / (float)total_char_count;
+			}
 		}
+		main->first_invalid_line = 0; //invalidate ALL
+		_validate_line_caches(main);
+		update();
 	}
-	update();
 }
```

Comment:
Changes in the `set_visible_characters` function are fine, and should fix #51884.

The rest seems wrong, it will break if there's more than one `ITEM_TEXT` (e.g. multiple fonts or image in the middle of text in the line), I do not see any issues with the original code, what was not working in it?

Reply:
```tx = tx.substr(0, l.char_offset + l.char_count + tx.length() - visible_characters);```

This line have errors. An example, lets say: (before the `Shape current paragraph` loop.)
```
tx = "abcdef"
l.char_offset = 0
l.char_count = 0 //initially all lines have the value 0 

visible_characters = 5
```

Expected result: "abcdef"
Real result: "a"       `//tx = tx.substr(0, 0 + 0 + 6 - 5)`

Reply:
As you mentioned, let me dig through the existing stuff a bit more to fix it without breaking the other stuff.

Reply:
I can try this:
`tx = tx.substr(0, tx.length() - (l.char_offset + l.char_count + tx.length() - visible_characters));`

But char_offset still have problems. Because counts `\n` as char as well.

Reply:
I will show two slightly different solutions. Coming up.

Reply:
@bruvzg 

https://github.com/requizm/godot/commit/a733cde5c8405d35e08b30d7a2633f1e7b4c1d5c

This can be solution too but I don't know if it will break other stuffs.

Reply:
Maybe we need to do it by checking that the `ITEM_NEXT` follows the `ITEM_NEWLINE`